// myfatoorah-reactnative/lib/module/MFEnums.js
export let MFEnvironment = /*#__PURE__*/function (MFEnvironment) {
  MFEnvironment["TEST"] = "TEST";
  MFEnvironment["LIVE"] = "LIVE";
  return MFEnvironment;
}({});
export let MFLanguage = /*#__PURE__*/function (MFLanguage) {
  MFLanguage["ARABIC"] = "ar";
  MFLanguage["ENGLISH"] = "en";
  return MFLanguage;
}({});
export let MFCurrencyISO = /*#__PURE__*/function (MFCurrencyISO) {
  MFCurrencyISO["KUWAIT_KWD"] = "KWD";
  MFCurrencyISO["SAUDIARABIA_SAR"] = "SAR";
  MFCurrencyISO["BAHRAIN_BHD"] = "BHD";
  MFCurrencyISO["UAE_AED"] = "AED";
  MFCurrencyISO["QATAR_QAR"] = "QAR";
  MFCurrencyISO["OMAN_MOR"] = "OMR";
  MFCurrencyISO["JORDAN_JOD"] = "JOD";
  MFCurrencyISO["UNITEDSTATES_USD"] = "USD";
  MFCurrencyISO["EURO_EUR"] = "EUR";
  return MFCurrencyISO;
}({});
export let MFCountry = /*#__PURE__*/function (MFCountry) {
  MFCountry["KUWAIT"] = "KWT";
  MFCountry["SAUDIARABIA"] = "SAU";
  MFCountry["BAHRAIN"] = "BHR";
  MFCountry["UNITEDARABEMIRATES"] = "ARE";
  MFCountry["QATAR"] = "QAT";
  MFCountry["OMAN"] = "OMN";
  MFCountry["JORDAN"] = "JOD";
  MFCountry["EGYPT"] = "EGY";
  return MFCountry;
}({});
export let MFMobileCountryCodeISO = /*#__PURE__*/function (MFMobileCountryCodeISO) {
  MFMobileCountryCodeISO["KUWAIT"] = "+965";
  MFMobileCountryCodeISO["SAUDIARABIA"] = "+966";
  MFMobileCountryCodeISO["BAHRAIN"] = "+973";
  MFMobileCountryCodeISO["UAE"] = "+971";
  MFMobileCountryCodeISO["QATAR"] = "+974";
  MFMobileCountryCodeISO["OMAN"] = "+968";
  MFMobileCountryCodeISO["JORDAN"] = "+962";
  MFMobileCountryCodeISO["EGYPT"] = "+20";
  return MFMobileCountryCodeISO;
}({});
export let MFNotificationOption = /*#__PURE__*/function (MFNotificationOption) {
  MFNotificationOption["ALL"] = "all";
  MFNotificationOption["EMAIL"] = "eml";
  MFNotificationOption["SMS"] = "sms";
  MFNotificationOption["LINK"] = "lnk";
  return MFNotificationOption;
}({});
export let MFKeyType = /*#__PURE__*/function (MFKeyType) {
  MFKeyType["INVOICEID"] = "InvoiceId";
  MFKeyType["PAYMENTID"] = "PaymentId";
  return MFKeyType;
}({});
export let MFFontWeight = /*#__PURE__*/function (MFFontWeight) {
  MFFontWeight["Thin"] = "Thin";
  MFFontWeight["ExtraLight"] = "ExtraLight";
  MFFontWeight["Light"] = "Light";
  MFFontWeight["Normal"] = "Normal";
  MFFontWeight["Medium"] = "Medium";
  MFFontWeight["SemiBold"] = "SemiBold";
  MFFontWeight["Bold"] = "Bold";
  MFFontWeight["ExtraBold"] = "ExtraBold";
  MFFontWeight["Heavy"] = "Heavy";
  return MFFontWeight;
}({});
export let MFFontFamily = /*#__PURE__*/function (MFFontFamily) {
  MFFontFamily["SansSerif"] = "SansSerif";
  MFFontFamily["Arial"] = "Arial";
  MFFontFamily["TimesNewRoman"] = "TimesNewRoman";
  MFFontFamily["Helvetica"] = "Helvetica";
  MFFontFamily["Tahoma"] = "Tahoma";
  MFFontFamily["Verdana"] = "Verdana";
  MFFontFamily["Monaco"] = "Monaco";
  MFFontFamily["Georgia"] = "Georgia";
  MFFontFamily["CourierNew"] = "CourierNew";
  MFFontFamily["Garamond"] = "Garamond";
  MFFontFamily["TrebuchetMS"] = "TrebuchetMS";
  return MFFontFamily;
}({});
export let MFRecurringType = /*#__PURE__*/function (MFRecurringType) {
  MFRecurringType["DAILY"] = "Daily";
  MFRecurringType["WEEKLY"] = "Weekly";
  MFRecurringType["MONTHLY"] = "Monthly";
  MFRecurringType["CUSTOM"] = "Custom";
  return MFRecurringType;
}({});
export let MFTokenType = /*#__PURE__*/function (MFTokenType) {
  MFTokenType["MFtoken"] = "mftoken";
  MFTokenType["GooglePay"] = "googlepay";
  return MFTokenType;
}({});
